src/app/fdctl/run/tiles/fd_sign.c

```diff
@@ -121,6 +121,32 @@ after_frag( void *             _ctx,
       fd_ed25519_sign( ctx->out[ in_idx ].data, ctx->_data, 130UL, ctx->public_key, ctx->private_key, ctx->sha512 );
       break;
     }
+    case FD_KEYGUARD_ROLE_GOSSIP: {
+      if( FD_UNLIKELY( !fd_keyguard_payload_authorize( ctx->_data, *opt_sz, FD_KEYGUARD_ROLE_GOSSIP ) ) ) {
+        FD_LOG_EMERG(( "fd_keyguard_payload_authorize failed %lu %u %u %u %u", *opt_sz, ctx->_data[0], ctx->_data[1], ctx->_data[2], ctx->_data[3] ));
+      }
+      if ( fd_keyguard_payload_matches_ping_msg( ctx->_data, *opt_sz ) ) {
+        /* Gossip tile sends the sh256 pre-image for ping/pong msgs. */
+        uchar hash[32];
+        fd_sha256_t sha[1];
+        fd_sha256_init( sha );
+        fd_sha256_append( sha, ctx->_data, *opt_sz );
+        fd_sha256_fini( sha, hash );
+
+        fd_ed25519_sign( ctx->out[ in_idx ].data, hash, 32UL, ctx->public_key, ctx->private_key, ctx->sha512 );
+      } else {
+        fd_ed25519_sign( ctx->out[ in_idx ].data, ctx->_data, *opt_sz, ctx->public_key, ctx->private_key, ctx->sha512 );
```

Comment:
? This branch should be a LOG_EMERG like the other ones

Reply:
For crds messages we sign the data directly, but for pings/pongs we sing the hash of the data. We have an FD_LOG_EMERG if the authorize fails which also contains this check